### DISCUSSION TITLE
Make it possible to put the language in the URL

### DISCUSSION THREAD

**Discussion by fiammybe** [ImpressCMS/impresscms#1031] [Ideas]
As described some time ago by Nachenko in [the forum]([https://www.impresscms.org/modules/iforum/viewtopic.php?topic_id=4020&forum=55&post_id=36841]), it would benefit multilanguage sites to have the language in the URL somewhere. It would be possible to make links directly to the correct language page, and it would help SEO as well.

This could have quite some ramifications if we want to do it correctly, so don't underestimate :-)

**Comment by skenow** [comment]
I thought the language was in the URL, at least as a parameter, if they switched to another language: lang=en, for example.

Google has made some recent admissions about the URL structure - https://www.searchenginejournal.com/google-click-depth-matters-seo-url-structure/256779/

**Comment by fiammybe** [comment]
OK, for google it's not a SEO enhancement, so i removed that from the title. 

You can change the language the system uses by using a parameter, true. That is the way we switch between different languages in the system. But a parameter is not the best way to differentiate different pages, and adding parameters the whole time is ugly (point 5 on https://moz.com/blog/15-seo-best-practices-for-structuring-urls)

**Comment by fiammybe** [comment]
Another reason to have this in the URL : [Google mentions](https://support.google.com/webmasters/answer/182192?hl=en&ref_topic=2370587) that geo-targeting is not possible based on parameters in the URL. So if I want to target the https://www.impresscms.org/de domain for Germany, I can't do it with https://www.impresscms.org/?lang=de

**Comment by MekDrop** [comment]
https://github.com/middlewares/negotiation#contentlanguage has possibility to detect language from browser or from url. Probably we could use such middleware for this problem solving.
